src/main/java/com/minecolonies/coremod/inventory/CraftingGUIBuilding.java

```diff
@@ -0,0 +1,350 @@
+        if (slotId >= 1 && slotId < 5)
+        {
+            // 1 is shift-click
+            if (mode == ClickType.PICKUP
+                    || mode == ClickType.PICKUP_ALL
+                    || mode == ClickType.SWAP)
+            {
+                Slot slot = this.inventorySlots.get(slotId);
+
+                ItemStack dropping = playerIn.inventory.getItemStack();
+
+                return handleSlotClick(slot, dropping);
+            }
+
+            return ItemStack.EMPTY;
+        }
+
+        return super.slotClick(slotId, clickedButton, mode, playerIn);
+    }
+
+    public ItemStack handleSlotClick(Slot slot, ItemStack stack)
+    {
+        if (stack.getCount() > 0)
+        {
+            ItemStack copy = stack.copy();
+            copy.setCount(1);
+            slot.putStack(copy);
+        }
+        else if (slot.getStack().getCount() > 0)
+        {
+            slot.putStack(ItemStack.EMPTY);
+        }
+
+        return slot.getStack().copy();
+    }
+
+    @Nullable
+    @Override
+    public ItemStack transferStackInSlot(final EntityPlayer playerIn, final int index)
+    {
+        if (index <= 5)
+        {
+            return ItemStack.EMPTY;
+        }
+
+        ItemStack itemstack = null;
+        final Slot slot = this.inventorySlots.get(index);
+
+        if (slot != null && slot.getHasStack())
+        {
+            final ItemStack itemstack1 = slot.getStack();
+            itemstack = itemstack1.copy();
+
+            if (index == 0)
+            {
+                if (!this.mergeItemStack(itemstack1, CRAFTING_SLOTS, TOTAL_SLOTS, true))
+                {
+                    return null;
+                }
+
+                slot.onSlotChange(itemstack1, itemstack);
+            }
+            else if (index >= CRAFTING_SLOTS && index < HOTBAR_START)
+            {
+                if (!this.mergeItemStack(itemstack1, HOTBAR_START, TOTAL_SLOTS, false))
+                {
+                    return null;
+                }
+            }
+            else if (index >= HOTBAR_START && index < TOTAL_SLOTS)
+            {
+                if (!this.mergeItemStack(itemstack1, CRAFTING_SLOTS, HOTBAR_START, false))
+                {
+                    return null;
+                }
+            }
+            else if (!this.mergeItemStack(itemstack1, CRAFTING_SLOTS, TOTAL_SLOTS, false))
+            {
+                return null;
+            }
+
+            if (itemstack1.getCount() == 0)
+            {
+                slot.putStack(ItemStackUtils.EMPTY);
+            }
+            else
+            {
+                slot.onSlotChanged();
+            }
+
+            if (itemstack1.getCount() == itemstack.getCount())
+            {
+                return null;
+            }
+
+            slot.onTake(playerIn, itemstack1);
+        }
+
+        return itemstack;
+    }
+
+    public List<Slot> getCraftingSlots()
+    {
+        return ImmutableList.of(getSlot(1), getSlot(2), getSlot(3), getSlot(4));
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Assign this magic number 3 to a well-named constant, and use the constant instead. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Assign this magic number 4 to a well-named constant, and use the constant instead. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)